### PR TITLE
test : Add unit tests for volume.go

### PR DESCRIPTION
**Pull request by supritai**
## Summary

Add a comprehensive test file [volume_test.go](vscode-webview://1fuo8f82k1llgdk7p4ee1gn5vb10a06367ttf2hs1lrm6na24re2/src/cloud-providers/libvirt/volume_test.go) for the libvirt cloud provider's volume utilities.

## Tests covered

TestNewDefVolume — verifies that newDefVolume produces correctly structured libvirtxml.StorageVolume objects (format, permissions, capacity unit) for various volume name styles.
TestNewDefVolumeFromXML — round-trip XML parsing tests including backing-store presence, alternate capacity units, and error paths (malformed / empty / wrong-element XML).
TestWaitForSuccess — table-driven tests for the retry/timeout helper, checking immediate success, N-retry success, timeout behaviour, and that the error message surfaces both the operation name and the last error.
TestVolumeCapacityBytesTable — edge-case coverage for volumeCapacityBytes, including zero inputs, backing-store vs requested size comparison, large safe values, and uint64 overflow detection.
## Notes

All tests are gated with //go:build cgo to match the build constraints of the package.
Test helpers setWaitTimers and assertVolumeDefShape are scoped to the test file to avoid polluting the package namespace.

Assisted-by: IBM Bob [noreply@ibm.com](mailto:noreply@ibm.com)
Signed-off-by: Suprit Iti [suprit.iti@ibm.com](mailto:suprit.iti@ibm.com)

### CODE REVIEW

File: src/cloud-providers/libvirt/volume_test.go [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+	defVolFormat = "qcow2"
+	// defVolPermMode is the expected permissions mode set by newDefVolume.
+	defVolPermMode = "644"
+	// defVolCapUnit is the expected capacity unit set by newDefVolume.
+	defVolCapUnit = "bytes"
+
+	// testVolumeCapacity is the 1 GiB capacity value (in bytes) used in XML fixtures.
+	// It equals 1073741824, the same value as bytesPerGiB.
+	testVolumeCapacity = uint64(1024 * 1024 * 1024)
+
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: Could we just use `bytesPerGiB` if this is already defined and in scope?

File: src/cloud-providers/libvirt/volume_test.go [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+			expectError: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			volumeDef, err := newDefVolumeFromXML(tt.xmlData)
+
+			if tt.expectError {
+				assertErrContains(t, err)
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: This doesn't make sense, do you have an error string to check here?

File: src/cloud-providers/libvirt/volume_test.go [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+	waitSleepInterval = interval
+	t.Cleanup(func() {
+		waitTimeout = orig
+		waitSleepInterval = origInterval
+	})
+}
+
+// assertErrContains asserts that err is non-nil and that its message contains
+// every substring in want.
+func assertErrContains(t *testing.T, err error, want ...string) {
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: Is it worth adding this for just a single test case, could you instead just concatenate them e.g.
```
assert.Contains(t, err.Error(), testOperation+": "+errPersistent)
```
, or worse case add a regex match or something like
```
pattern := regexp.QuoteMeta(testOperation) + ".*" + regexp.QuoteMeta(errPersistent)
assert.Regexp(t, pattern, err.Error())
```
?

File: src/cloud-providers/libvirt/volume_test.go [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"libvirt.org/go/libvirtxml"
+)
+
+// Test-scoped constants for fixed domain values used across multiple test cases.
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: This looks like AI-slop? I think we know what constant in a test case are for

File: src/cloud-providers/libvirt/volume_test.go [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+// Note: testVolumeName is declared in image_test.go and shared across the package.
+const (
+	// defVolFormat is the expected qcow2 format set by newDefVolume.
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: I'm unsure of the value of any of these comments as well, which mostly just restate the names

File: src/cloud-providers/libvirt/volume_test.go [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+				tt.volumeName, tt.wantFormat, tt.wantPermMode,
+				tt.wantCapacityUnit, tt.wantCapacityValue)
+		})
+	}
+}
+
+func TestNewDefVolumeFromXML(t *testing.T) {
+	// xmlCapacity is the string representation of testVolumeCapacity (1 GiB in
+	// bytes) embedded in XML fixtures. Must stay in sync with testVolumeCapacity.
+	const xmlCapacity = "1073741824"
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: Can you use `strconv.FormatUint(bytesPerGiB, 10)` please as it's much less brittle

File: src/cloud-providers/libvirt/volume_test.go [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+		wantCapacityUnit  string
+		wantCapacityValue uint64
+	}{
+		{
+			name:              "simple name",
+			volumeName:        "test-vol",
+			wantFormat:        defVolFormat,
+			wantPermMode:      defVolPermMode,
+			wantCapacityUnit:  defVolCapUnit,
+			wantCapacityValue: 1,
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: These are identical in every test case, why are the needed in the test data struct?

File: src/cloud-providers/libvirt/volume_test.go [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+	}
+}
+
+// TestBytesPerGiBConstant guards that bytesPerGiB equals exactly 1 GiB in
+// bytes. The two representations (power-of-two expression and decimal) must
+// both match so a future accidental redefinition is caught immediately.
+func TestBytesPerGiBConstant(t *testing.T) {
+	assert.Equal(t, uint64(1024*1024*1024), bytesPerGiB, "bytesPerGiB should equal 2^30")
+	assert.Equal(t, uint64(1073741824), bytesPerGiB, "bytesPerGiB should equal 1073741824")
+}
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: What is the reason for this - it's testing that golang can do maths or something?

File: src/cloud-providers/libvirt/volume_test.go (hunk 1 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: sure

File: src/cloud-providers/libvirt/volume_test.go (hunk 4 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: yes, will remove all constant comments

File: src/cloud-providers/libvirt/volume_test.go (hunk 3 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: yes, concatenate really helps. Thanks for suggesting it

File: src/cloud-providers/libvirt/volume_test.go (hunk 7 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: `ssertVolumeDefShape` now references `defVolFormat` and `defVolPermMode `directly, and the table rows are collapsed to two

File: src/cloud-providers/libvirt/volume_test.go (hunk 6 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: sure

File: src/cloud-providers/libvirt/volume_test.go (hunk 8 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: Removed. Both assertions were comparing compile-time constants, so the test could never fail in practice. Thanks for pointing it.

File: src/cloud-providers/libvirt/volume_test.go [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+			expectError: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			volumeDef, err := newDefVolumeFromXML(tt.xmlData)
+
+			if tt.expectError {
+				assert.Error(t, err)
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: Can we swap these into `assert.ErrorContains(t, err, <message>)` so we are actually checking what went wrong?

File: src/cloud-providers/libvirt/volume_test.go [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+		// zero value means "skip this assertion" for rows that use checkFunc.
+		wantName     string
+		wantCapacity uint64
+		wantFormat   string
+		checkFunc    func(*testing.T, libvirtxml.StorageVolume)
+	}{
+		{
+			name: "valid volume XML",
+			xmlData: `<volume>
+				<name>test-volume</name>
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: Can we use `testVolumeName` in the XML so the tests that check it are less brittle please

File: src/cloud-providers/libvirt/volume_test.go [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+					<format type="qcow2"/>
+					<permissions>
+						<mode>0644</mode>
+					</permissions>
+				</target>
+			</volume>`,
+			expectError:  false,
+			wantName:     testVolumeName,
+			wantCapacity: bytesPerGiB,
+			wantFormat:   defVolFormat,
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: This current pattern of this with wants and the rest with checkFunc feels messy to me, can you update this to:
```
			checkFunc: func(t *testing.T, vol libvirtxml.StorageVolume) {
				assert.Equal(t, "test-volume", vol.Name)
				assert.Equal(t, bytesPerGiB, vol.Capacity.Value)
				assert.Equal(t, defVolFormat, vol.Target.Format.Type)
			},
```
to help with this?

File: src/cloud-providers/libvirt/volume_test.go [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+				<backingStore>
+					<path>/var/lib/libvirt/images/base.qcow2</path>
+					<format type="qcow2"/>
+				</backingStore>
+			</volume>`,
+			expectError: false,
+			checkFunc: func(t *testing.T, vol libvirtxml.StorageVolume) {
+				assert.NotNil(t, vol.BackingStore)
+				assert.Contains(t, vol.BackingStore.Path, "base.qcow2")
+			},
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go:122]: I guess this would become:
```
			checkFunc: func(t *testing.T, vol libvirtxml.StorageVolume) {
				assert.Equal(t, "test-volume", vol.Name)
				assert.Equal(t, bytesPerGiB, vol.Capacity.Value)
				assert.NotNil(t, vol.BackingStore)
				assert.Contains(t, vol.BackingStore.Path, "base.qcow2")
			},
```

File: src/cloud-providers/libvirt/volume_test.go [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result, err := volumeCapacityBytes(tt.volSizeGiB, tt.backingCapacityBytes)
+
+			if tt.wantErr {
+				assert.Error(t, err)
+				assert.Contains(t, err.Error(), tt.errContains)
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: Please update this with:
```
assert.ErrorContains(t, err, tt.errContains)
```

File: src/cloud-providers/libvirt/volume_test.go [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+					return errors.New(errPersistent)
+				}
+				if callCount <= tt.failCount {
+					return fmt.Errorf("attempt %d failed", callCount)
+				}
+				return nil
+			})
+
+			if tt.wantErr {
+				assert.Contains(t, err.Error(), testOperation+": "+errPersistent)
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: This can be `assert.ErrorContains(t, err, testOperation+": "+errPersistent)` too

File: src/cloud-providers/libvirt/volume_test.go (hunk 9 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: sure

File: src/cloud-providers/libvirt/volume_test.go (hunk 10 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: Done — all three XML fixtures now use testVolumeName via string concatenation, same pattern as xmlCapacity.

File: src/cloud-providers/libvirt/volume_test.go (hunk 11 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: Done — moved the assertions from wantName/wantCapacity/wantFormat fields into a checkFunc.

File: src/cloud-providers/libvirt/volume_test.go (hunk 12 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go:122]: yes

File: src/cloud-providers/libvirt/volume_test.go (hunk 13 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: sure

File: src/cloud-providers/libvirt/volume_test.go (hunk 14 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: yes

**Comment by stevenhorsman** [comment]
@supritai - why have you requested a re-review when you haven't made any changes to the code since my last review? Also please rebase properly and remove the merge commit. Thanks

File: src/cloud-providers/libvirt/volume_test.go [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+}
+
+func TestVolumeCapacityBytesTable(t *testing.T) {
+	tests := []struct {
+		name                 string
+		volSizeGiB           uint64
+		backingCapacityBytes uint64
+		wantBytes            uint64
+		wantErr              bool
+		errContains          string
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go]: Could we have these as a single value like above for consistency please:
```
wantErr   string // non-empty: expect error containing this substring
```

File: src/cloud-providers/libvirt/volume_test.go [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+		},
+		{
+			name:    "malformed XML — unclosed tag",
+			xmlData: `<volume><name>test</name>`,
+			wantErr: "unexpected EOF",
+		},
+		{
+			name:    "empty XML",
+			xmlData: ``,
+			wantErr: "EOF",
```
Comment (stevenhorsman) [inline src/cloud-providers/libvirt/volume_test.go:151]: Is this another unexpected EOF error like the one above?

**Comment by supritai** [comment]
> @supritai - why have you requested a re-review when you haven't made any changes to the code since my last review? Also please rebase properly and remove the merge commit. Thanks

Oh.. Yesterday push was failed. Now I observed. Sorry about it

File: src/cloud-providers/libvirt/volume_test.go (hunk 15 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go]: sure

File: src/cloud-providers/libvirt/volume_test.go (hunk 16 above)
Comment (supritai) [reply src/cloud-providers/libvirt/volume_test.go:151]: They're actually different errors — empty XML returns a plain io.EOF ("EOF"), while the unclosed tag returns an *xml.SyntaxError with message "XML syntax error on line 1: unexpected EOF".

Comment (stevenhorsman) [reply src/cloud-providers/libvirt/volume_test.go:151]: Cool - thanks for confirming

**Review by stevenhorsman — APPROVED** [review]
Looks okay to me. Thanks

**Comment by supritai** [comment]
@mkulke @mythi Please review.

**Review by mkulke — APPROVED** [review]
